Formula/llvm.rb

```diff
@@ -62,15 +60,14 @@ def install
     ]
     runtimes = %w[
       compiler-rt
-      libcxx
-      libcxxabi
       libunwind
+      openmp
     ]
-    if OS.mac?
-      runtimes << "openmp"
-    else
-      projects << "openmp"
-    end
+
+    # Linking with these causes you subtle problems on macOS,
+    # since everything loads `/usr/lib/libc++.1.dylib`.
+    # https://github.com/Homebrew/homebrew-core/issues/96915
+    runtimes += %w[libcxx libcxxabi] if OS.linux?
```

Comment:
Thoughts on this, @Bo98?

On the one hand, I feel like it might be a bit of an overreaction. On the other, we already prefer linking with `/usr/lib/libc++.1.dylib` anyway, so it's not like we have much use for these...

Reply:
Also, I can move this up into the `else` block above. I'll do so on the next run.

Reply:
I agree that these need to go. Or if they do ship then they cannot live in `lib` anymore.

Ideally the C++ headers would stay however, since I think those are necessary for clang++ usage on < 10.15.3 SDK and shouldn't affect anything on newer SDKs.

Reply:
Alternative is to symlink the C++ header directory to the CLT header directory.

Reply:
I'm still a little sceptical that this is the right solution, though, since there's been lots of use of the bundled `libc++` for a long time, even in other formulae (despite our efforts to avoid it). So I don't think the issue is that you're linking with two versions of `libc++` -- that's been happening for a _long_ time.

Reply:
> since there's been lots of use of the bundled `libc++`, even in other formulae (despite our efforts to avoid it)

Which ones? I know v8 tries to, but it also carefully controls all of its dependencies to make sure everything uses the same `libc++`.

Linking LLVM `libc++` should be opt-in - not a default. We have several hacks in formulae specifically trying to work around that mess.

Reply:
> Which ones?

Check out the list of linkage failures:
- c2rust
- castxml
- ccls
- chapel (though this one has reported problems)
- clazy
- coin3d
- crystal
- csound
- enzyme
- faust
- halide
- include-what-you-use
- ispc
- ldc
- nvc
- odin
- pocl
- pyside
- pyside@2
- rtags
- spirv-llvm-translator
- standardese
- symengine

> I know v8 tries to, but it also carefully controls all of its dependencies to make sure everything uses the same `libc++`.

The extent that you can do that is limited, since `libSystem.B.dylib` also loads `/usr/lib/libc++.1.dylib`. (They depend on each other. System libraries are weird. That's part of why `ld` has to build these hacks like "upward" library dependencies, I think.)

Reply:
Also, Catalina failed because the compiler couldn't find the C++ headers. We probably need to at least ship the headers or symlink the CLT ones into the right place.

I'm not keen on shipping the headers alone -- if we are convinced that mixing `libc++` versions is a no-no (or at least, enough to not ship `libc++`), then we probably shouldn't be allowing mixing upstream LLVM headers with the system `libc++`.

Reply:
> we probably shouldn't be allowing mixing upstream LLVM headers with the system `libc++`.

Typically yeah, mismatching headers and libraries is not good in general. But note that libc++ ships with macOS availability macros that issue compile time errors if you use a newer functions on an older macOS. This matches how much of how Apple frameworks work - so you can use the latest headers and still build for older OS versions and their older runtime libraries. It's also worth noting a significant portion of C++'s STL is header-only anyway, so using newer headers can still give benefits even if the library is old.

Shipping newer C++ headers is what Apple have done for years. Even after the move to the SDK (which AFAIK was done for the benefit of Swift's C++ interoperability rather than anything libc++ related - hence why we only support Catalina and later for the Swift formula), Xcode only ships with the latest SDK unlike the CLT so you are still using newer headers than libraries in those scenarios. Using a newer SDK is safe, as long as you're not using low level system headers like `curl`.

MacPorts even shipped `LIBCXX_INSTALL_LIBRARY=OFF` for years (https://lists.macports.org/pipermail/macports-dev/2014-May/026820.html). They managed to turn it on safely, but they don't ship shared LLVM libraries (libLLVM) and linking against the static libraries won't start adding rpaths to the LLVM lib directory, so accidentally linking against LLVM's libc++ doesn't really happen.

I think a setup to strongly consider is installing libcxx like before but with `LIBCXX_INSTALL_LIBRARY_DIR` set to `lib/c++` (on macOS) or something. This means having to opt-in pass `-L$HOMEBREW_PREFIX/opt/llvm/lib/c++`, which I think is much safer and more explicit. Note using LLVM's libc++ already typically involves custom compiler flags: you would want to pass `-D_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS` otherwise the availability macros are just going to stop you from using any new feature anyway. If a software is intentionally wanting to use LLVM's libc++, we should know about it.

Reply:
> I think a setup to strongly consider is installing libcxx like before but with `LIBCXX_INSTALL_LIBRARY_DIR` set to `lib/c++` (on macOS) or something. This means having to opt-in pass `-L$HOMEBREW_PREFIX/opt/llvm/lib/c++`, which I think is much safer and more explicit. Note using LLVM's libc++ already typically involves custom compiler flags: you would want to pass `-D_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS` otherwise the availability macros are just going to stop you from using any new feature anyway by applying logic based on the system libc++. If a software is intentionally wanting to use LLVM's libc++, we should know about it.

I like this.

I wonder if we can consider making `llvm` not-keg-only with this, since, AFAICT, none of the remaining libraries shadow ones with install names starting with `/usr/lib`.

You do shadow `clang` and `clang++`, but these generally have a stable command-line API (and is typically the desired behaviour for most users who do `brew install llvm`).

Reply:
Trying this.

Though, the `scan-build` test is failing, even after trying to point it to `bin/"clang"`. Seems to be `libc++`-related, since it didn't fail when I removed `libc++`.

Reply:
Moving libc++ might end up fixing that, we'll see.

Reply:
Oh wait, didn't notice it's already failing with the change. Does LLVM 14 fail with a similar absolute clang reference?

Reply:
> Does LLVM 14 fail with a similar absolute clang reference?

It does not. It works with an unqualified `clang++` or a `bin/"clang++"`.

Reply:
Error looks related to `libc++` headers.